### DISCUSSION TITLE
Exploitation and abuse of server resources (maybe)

### DISCUSSION THREAD

**Discussion by promitheas-nikou** [cc-tweaked/CC-Tweaked#1120] [Q&A] [answered]
If computers have their own memory, can execute arbitrary (Lua) code, and can have access to the internet through the http api, what prevents someone from connecting to a server and placing down many computers and have them, for example, mine bitcoin for him (in real life), while using the server's resources?

**Comment by SirEdvin** [comment]
yielding, if I remember correctly. 

More efficient that this, you can just setup lag machine with hopper or just use another mods, like mekanism

**Reply by promitheas-nikou** [reply]
Yes, but unlike all those other things that also consume server resources, people will actually benefit (not really, computers are really slow, but whatever) from this

**Reply by SirEdvin** [reply]
Not really, on my opinion. Due to yielding and other specific limitations of CC:T computers profit will be nearly zero.

You also will need to implemented mining script on lua that will have correct yielding inside code (because without yielding computer will be killed after 5 seconds or less.

**Comment by SquidDev** [comment, marked as answer]
There's a couple of reasons:
 - All computers are run on a fixed number of threads (defaults to 1). So placing down two computers won't use any more CPU than one computer will.
 - Computers are _incredibly_ slow (>50 times slower than _single threaded_ C, many times slower again than the dedicated GPUs/ASICs miners use). I suspect even if you ran a miner on every single modded MC server out there, you'd still never get a single coin.

Also worth noting that CC has some pretty good monitoring tools (see the `/computercraft track` command), so you can keep an eye on computers using a lot of CPU time and stop them if you think they're doing something dodgy.

**Reply by promitheas-nikou** [reply]
But if all computers in a server run on a fixed number of threads, what are we supposed to do when there's just a lot of players in the server that need many computers? Do we have to increase it manually?

**Reply by SquidDev** [reply]
Much like a normal computer, CC runs multiple computers on a single thread, and can pre-emptively pause them and run others if needed (much like how a normal computer can run processes on different cores).

Of course, if you've got a lot of computers, things will run a little slow. At that point you can bump the thread count, though I've rarely had to do that - you can easily 50+ active computers on a single thread without issue.

**Reply by promitheas-nikou** [reply]
Still though, if the current problem is that they're really slow, we can use them for cloud storage ( :
I could just put a single computer and 50 disk drives and connect using the http api to some server and store its files or whatever
Is there a limit to the total memory allocated to computers?

**Reply by SquidDev** [reply]
Each computer can store 1Mb of data, so you'd need to place about 20k to match the free tier of Google Drive.

Worth noting there is a config option to disable the HTTP API if you're concerned about players exploiting it.
